Fixed #34262 -- Added support for AnyValue for MySQL, Oracle, and Postgresql 16+.

ontowhee opened the pull request:
#### Trac ticket number

ticket-34262

#### Branch description

Added AnyValue() aggregation function to allow for aggregations involving columns that are not determined to be functionally dependent on the expressions used in a group by clause.

#### Checklist
- [x] This PR targets the `main` branch.
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [ ] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

tim-schilling commented on docs/ref/models/querysets.txt:
```diff
@@ -3930,6 +3930,54 @@ when the queryset (or grouping) contains no entries.
+
+        If you're using MySQL and ``sql_mode`` option contains ``ONLY_FULL_GROUP_BY``,
+        it may be necessary to use ``AnyValue`` if an aggregation includes a
+        mix of aggregated and non-aggregated functions.
```
Would it be helpful to link people to the relevant DB documentation here? It looks like we do link to db provider documentation, but I'm not sure how prevalent the practice is.

charettes replied:
I agree, it might be useful to link to either of

- https://dev.mysql.com/doc/refman/8.4/en/group-by-handling.html
- https://dev.mysql.com/doc/refman/8.4/en/miscellaneous-functions.html#function_any-value

ontowhee replied:
Thanks for the suggestions. I've updated the wording and linked to the group by handling documentation.

tim-schilling commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,55 @@ aggregate that author count, referencing the annotation field:
+
+The database is not able to determine that the ``greatest_pages`` used in
+``pages_per_author`` is functionally dependent on the column in the group by
+clause. In MySQL, this will happen if the ``sql_mode`` contains
+``ONLY_FULL_GROUP_BY``, and it will raise the
+``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
```
Reading this as someone who doesn't know much about the feature, do I need to be worried about this error with non-mysql databases? If the answer is no, do we still want to mention the use case of why we should use `AnyValue` with non-mysql databases?

charettes replied:
> Reading this as someone who doesn't know much about the feature, do I need to be worried about this error with non-mysql databases?

I think we don't and share your sentiment that the documentation should be adjusted accordingly.

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1622,19 +1623,110 @@ def test_group_by_exists_annotation(self):
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value_filter(self):
+        books_qs = (
+            Book.objects.annotate(greatest_pages=Greatest("pages", Value(600)))
+            .values("greatest_pages")
```
```suggestion
            Book.objects.values(greatest_pages=Greatest("pages", Value(600)))
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1622,19 +1623,110 @@ def test_group_by_exists_annotation(self):
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value_over(self):
+        books_qs = (
+            Book.objects.annotate(
+                greatest_pages=Greatest("pages", 600),
+            )
+            .annotate(
+                pubdate_year=Window(
+                    AnyValue("pubdate__year"),
+                    partition_by="greatest_pages",
+                ),
+            )
+            .values_list("pubdate_year", flat=True)
+        )
```
```suggestion
        books_qs = Book.objects.annotate(
            greatest_pages=Greatest("pages", 600),
            pubdate_year=Window(
                AnyValue("pubdate__year"),
                partition_by="greatest_pages",
            ),
        ).values_list("pubdate_year", flat=True)
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1622,19 +1623,110 @@ def test_group_by_exists_annotation(self):
+
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value_filter_over(self):
+        books_qs = (
+            Book.objects.annotate(
+                greatest_pages=Greatest("pages", 600),
+            )
+            .annotate(
+                pubdate_year=Window(
+                    AnyValue("pubdate__year", filter=Q(rating__lt=4.5)),
+                    partition_by="greatest_pages",
+                ),
+            )
+            .values_list("pubdate_year", flat=True)
+        )
```
```suggestion
        books_qs = Book.objects.annotate(
            greatest_pages=Greatest("pages", 600),
            pubdate_year=Window(
                AnyValue("pubdate__year", filter=Q(rating__lt=4.5)),
                partition_by="greatest_pages",
            ),
        ).values_list("pubdate_year", flat=True)
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1622,19 +1623,110 @@ def test_group_by_exists_annotation(self):
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value_aggregate_clause(self):
+        books_qs = (
+            Book.objects.annotate(greatest_pages=Greatest("pages", 600))
+            .values("greatest_pages")
```
```suggestion
            Book.objects.values(greatest_pages=Greatest("pages", 600))
```

charettes reviewed (commented):
Thanks for the initial patch so far @ontowhee, this is impressive!

[SQLite support](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.create_aggregate) can be added with the following patch

```diff
diff --git a/django/db/backends/sqlite3/_functions.py b/django/db/backends/sqlite3/_functions.py
index 6d07d3d78b..01c9249709 100644
--- a/django/db/backends/sqlite3/_functions.py
+++ b/django/db/backends/sqlite3/_functions.py
@@ -80,6 +80,8 @@ def register(connection):
     connection.create_aggregate("STDDEV_SAMP", 1, StdDevSamp)
     connection.create_aggregate("VAR_POP", 1, VarPop)
     connection.create_aggregate("VAR_SAMP", 1, VarSamp)
+    connection.create_aggregate("ANY_VALUE", 1, AnyValue)
+    connection.create_window_function("ANY_VALUE", 1, AnyValue)
     # Some math functions are enabled by default in SQLite 3.35+.
     sql = "select sqlite_compileoption_used('ENABLE_MATH_FUNCTIONS')"
     if not connection.execute(sql).fetchone()[0]:
@@ -513,3 +515,21 @@ class VarPop(ListAggregate):

 class VarSamp(ListAggregate):
     finalize = statistics.variance
+
+
+class AnyValue(ListAggregate):
+    def value(self):
+        if self:
+            return self[0]
+        return None
+
+    def inverse(self, value):
+        try:
+            self.remove(value)
+        except ValueError:
+            pass
+
+    def finalize(self):
+        if self:
+            return self[0]
+        return None
diff --git a/django/db/backends/sqlite3/features.py b/django/db/backends/sqlite3/features.py
index 60893561df..78e828893f 100644
--- a/django/db/backends/sqlite3/features.py
+++ b/django/db/backends/sqlite3/features.py
@@ -34,6 +34,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_frame_range_fixed_distance = True
     supports_frame_exclusion = True
     supports_aggregate_filter_clause = True
+    supports_any_value = True
     order_by_nulls_first = True
     supports_json_field_contains = False
     supports_update_conflicts = True
```

But it segfaults on `aggregation.tests.AggregateTestCase.test_any_value_filter_over` even on the latest version of SQLite which I'm unsure why.

The generated SQL is

```sql
SELECT 
  ANY_VALUE(
    django_date_extract(
      year, "aggregation_book"."pubdate"
    )
  ) FILTER (
    WHERE 
      (
        "aggregation_book"."rating" < 4.5
      )
  ) OVER (
    PARTITION BY MAX("aggregation_book"."pages", 600)
  ) AS "pubdate_year" 
FROM 
  "aggregation_book
```

which I assumed was crashing because it's using nested user-defined function calls (notice `django_date_extract` usage) but `test_any_value_filter_over` does it without any issues. I'm pretty sure this is a SQLite bug with user-defined window functions making use of `FILTER (WHERE ...)` because when I disable the `supports_aggregate_filter_clause` feature flag it doesn't crash anymore.

Maybe we could world around that by having `ANY_VALUE` when used as a window function on SQLite turned into the native `first_value` which is equivalent to the proposed implementation here.

ontowhee commented:
@charettes Thank you for the patch for sqlite. I didn't know about python sqlite3 module, or about user defined functions, or FIRST_VALUE.

> Maybe we could world around that by having `ANY_VALUE` when used as a window function on SQLite turned into the native `first_value` which is equivalent to the proposed implementation here.

Interesting idea, however, if I understand correctly, this might not work because FILTER can only be used with aggregate window functions, and FIRST_VALUE is not an aggregate function.

I was running a simple query like this,
```
SELECT id, first_name, dob, first_value(first_name)
FILTER(WHERE dob > '2024-01-01')
OVER()
from person;
```
in sqlite and got the error `FILTER clause may only be used with aggregate window functions`.

---

I've updated the code with the following changes so it does not support window function:
- Added your patch for sqlite support, but without `connection.create_window_function()`, and without `inverse()` method.
- Added `AnyValue.window_compatible = False`.
- Removed `test_any_value_over` and `test_any_value_filter_over`.

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -3934,6 +3934,54 @@ when the queryset (or grouping) contains no entries.
+        ...     "headline",
+        ...     filter=Q(rating__gt=4.5),
+        ... )
+        >>> Entry.objects.annotate(
+        ...     pub_year=F("pub_date__year"),
+        ... ).values("pub_year").annotate(
```
```suggestion
        >>> Entry.objects.values(
        ...     pub_year=F("pub_date__year"),
        ... ).annotate(
```

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -3934,6 +3934,54 @@ when the queryset (or grouping) contains no entries.
+        >>> # from that year.
+        >>> from django.db.models import AnyValue, Avg, F, Q
+        >>> sample_headline = AnyValue("headline")
+        >>> Entry.objects.annotate(
+        ...     pub_year=F("pub_date__year"),
+        ... ).values("pub_year").annotate(
```
```suggestion
        >>> Entry.objects.values(
        ...     pub_year=F("pub_date__year"),
        ... ).annotate(
```

charettes commented on django/db/backends/mysql/features.py:
```diff
@@ -302,3 +293,9 @@ def allows_group_by_selected_pks(self):
+    def supports_any_value(self):
+        if self.connection.mysql_is_mariadb:
+            return False
+        return "ONLY_FULL_GROUP_BY" in self.connection.sql_mode
```
Is it the case that MySQL doesn't support it when `ONLY_FULL_GROUP_BY` is disabled? I thought it was a noop in this case which means we could unconditionally allow its usage on MySQL.

ontowhee replied:
You're right, it should be supported unconditionally! I've removed the condition.

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -3934,6 +3934,54 @@ when the queryset (or grouping) contains no entries.
+
+    .. note::
+
+        Supported by MySQL, Oracle, and PostgreSQL 16+.
```
```suggestion
        Supported on SQLite, MySQL, Oracle, and PostgreSQL 16+.
```

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -3934,6 +3934,54 @@ when the queryset (or grouping) contains no entries.
+
+        Supported by MySQL, Oracle, and PostgreSQL 16+.
+
+        If you're using MySQL and ``sql_mode`` option contains ``ONLY_FULL_GROUP_BY``,
```
```suggestion
        When the ``ONLY_FULL_GROUP_BY`` SQL mode is enabled on MySQL
```

charettes commented on docs/releases/6.0.txt:
```diff
@@ -199,6 +199,9 @@ Models
   :ref:`a forced update <ref-models-force-insert>` results in no affected rows,
   instead of a generic :exc:`django.db.DatabaseError`.
 
+* Added support for :class:`~django.db.models.AnyValue` for MySQL, Oracle, and
```
```suggestion
* Added support for :class:`~django.db.models.AnyValue` for SQLite, MySQL, Oracle, and
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,55 @@ aggregate that author count, referencing the annotation field:
 
+When you are generating an aggregate by using ``.values()`` clause to create the grouping
+for the annotation, it may be necessary to apply :class:`~django.db.models.AnyValue`
+when the annotation includes a mix of aggregated and non-aggregated functions. This
```
Are we aware that this might be a problem on other backends than MySQL?

ontowhee replied:
It doesn't seem to be a problem for other backends as far as I can tell when removing AnyValue from `test_any_value_aggregate_clause`. I've updated this to make it clear that this is a problem for MySQL.

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,55 @@ aggregate that author count, referencing the annotation field:
+When you are generating an aggregate by using ``.values()`` clause to create the grouping
+for the annotation, it may be necessary to apply :class:`~django.db.models.AnyValue`
+when the annotation includes a mix of aggregated and non-aggregated functions. This
+happens when the database is not be able to determine the functional dependency of the
+columns in the non-aggregated functions to the grouping column.
```
```suggestion
happens when the database is not able to determine the functional dependency of the
columns in the non-aggregated functions to the grouping column.
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,55 @@ aggregate that author count, referencing the annotation field:
+.. code-block:: pycon
+
+    >>> from django.db.models import F, Count, Greatest
+    >>> Book.objects.annotate(greatest_pages=Greatest("pages", 600)).values(
+    ...     "greatest_pages"
+    ... ).annotate(
```
```suggestion
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600)
    ... ).annotate(
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,55 @@ aggregate that author count, referencing the annotation field:
+.. code-block:: pycon
+
+    >>> from django.db.models import F, Count, Greatest
+    >>> Book.objects.annotate(greatest_pages=Greatest("pages", 600)).values(
+    ...     "greatest_pages"
+    ... ).annotate(
```
```suggestion
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600),
    ... ).annotate(
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1662,19 +1663,71 @@ def test_group_by_exists_annotation(self):
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value(self):
+        books_qs = (
+            Book.objects.values(greatest_pages=Greatest("pages", Value(600)))
```
```suggestion
            Book.objects.values(greatest_pages=Greatest("pages", 600))
```

charettes commented on tests/aggregation/tests.py:
```diff
@@ -1662,19 +1663,71 @@ def test_group_by_exists_annotation(self):
+    @skipUnlessDBFeature("supports_any_value")
+    def test_any_value_filter(self):
+        books_qs = (
+            Book.objects.values(greatest_pages=Greatest("pages", Value(600)))
```
```suggestion
            Book.objects.values(greatest_pages=Greatest("pages", 600))
```

charettes commented:
buildbot, test on oracle.

charettes commented on django/db/backends/mysql/features.py:
```diff
@@ -116,16 +116,7 @@ def django_test_skips(self):
-                    },
-                }
-            )
+
```
Very minor but we try to avoid unnecessary changes like these as they blur the blame trace. No need to push an amended commit just for that, the PR will need to be rebased anyway for an eventual merge so it can be dealt with eventually.

ontowhee replied:
Oh, I must have missed this extra line when I removed the skipped tests for the window functions! I can clean it up.

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,54 @@ aggregate that author count, referencing the annotation field:
+``pages_per_author`` is functionally dependent on the column in the group by
+clause, and it will raise the
+``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
+To avoid the ``OperationalError``, you can wrap the non-aggregated function
```
```suggestion
To avoid the ``OperationalError``, you can wrap the non-aggregate expression
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,54 @@ aggregate that author count, referencing the annotation field:
+The column in the group by clause is the expression for ``greatest_pages``.
+MySQL is not able to determine that the ``greatest_pages`` referenced in
+``pages_per_author`` is functionally dependent on the column in the group by
+clause, and it will raise the
+``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
```
I wonder if we should use _might_ here as MySQL might get smarter about that in the future

```suggestion
clause, and it might raise a
``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,54 @@ aggregate that author count, referencing the annotation field:
+When you are generating an aggregate by using ``.values()`` clause to create the
+grouping for the annotation on MySQL, and the ``ONLY_FULL_GROUP_BY`` SQL mode
+is enabled, it may be necessary to apply :class:`~django.db.models.AnyValue` if
+the annotation includes a mix of aggregated and non-aggregated functions.
```
I think we usually use `aggregate` and not `aggregated` to refer to such functions.

```suggestion
the annotation includes a mix of aggregate and non-aggregate expressions.
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,54 @@ aggregate that author count, referencing the annotation field:
+
+This creates groups of books based on the SQL column ``GREATEST(pages, 600)``.
+One unique group consists of books with 600 pages or less, and other unique
+groups will consist of books with the same pages. The ``pages_per_author``
+involves a mix of aggregated and non-aggregated functions, because
+``greatest_page`` is a non-aggregated function, whereas the ``num_authors`` is
+an aggregated function.
```
```suggestion
groups will consist of books with the same pages. The ``pages_per_author``
annotation is composed of aggregate and non-aggregate expressions,
``num_authors`` is an aggregate expression while ``greatest_page`` isn't.
```

charettes commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,54 @@ aggregate that author count, referencing the annotation field:
+``AnyValue`` is useful for working around columns in the select list that
+involve non-aggregated functions or complex expressions that are not determined
+by the database to be functionally dependent on the columns in the grouping clause.
+It is an optimized alternative to the ``Min`` and ``Max`` aggregate functions.
```
Not sure if this part is relevant as `ANY_VALUE` is allowed to pick any value even if `ORDER BY` is specified AFAIK. In other words, the fact some backends will honor the `ORDER BY` clause and pick the first item is an implementation detail?

charettes replied:
Sorry for the ambiguity, I meant to target only the line about the optimization.

tim-schilling reviewed (approved):
Approving to mark my earlier documentation concerns as resolved. Thanks @ontowhee!

ontowhee commented:
Thank you both! I learned so much from working on this. @tim-schilling Thanks for hosting Cosmic Contributors. It gave me the boost of confidence I needed along the way.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce reviewed (commented):
Thank you @ontowhee - this looks great! 
I have initial comments :+1:

ontowhee commented on django/db/models/aggregates.py:
```diff
@@ -229,6 +230,20 @@ def _get_repr_options(self):
+    def as_sql(self, compiler, connection, **extra_context):
+        if not connection.features.supports_any_value:
+            raise NotSupportedError(
+                "ANY_VALUE is not supported on this database backend."
```
Good catch! I've added the test.

ontowhee commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,52 @@ aggregate that author count, referencing the annotation field:
     >>> Book.objects.annotate(num_authors=Count("authors")).aggregate(Avg("num_authors"))
     {'num_authors__avg': 1.66}
 
+When you are generating an aggregate by using ``.values()`` clause to create the
```
I gave this a sub heading and linked to it in the AnyValue admonition. I wasn't entirely sure of the wording though. Does it seem all right?

sarahboyce replied:
Yes :star:

ontowhee commented on docs/ref/models/querysets.txt:
```diff
@@ -3943,6 +3943,60 @@ when the queryset (or grouping) contains no entries.
+
+    Supported on SQLite, MySQL, Oracle, and PostgreSQL 16+.
+
+    .. admonition:: MySQL with ``ONLY_FULL_GROUP_BY`` enabled
```
Replaced note with admonition and a title.

ontowhee commented on docs/ref/models/querysets.txt:
```diff
@@ -3943,6 +3943,60 @@ when the queryset (or grouping) contains no entries.
+        aggregate and non-aggregate functions. Using ``AnyValue`` allows the
+        non-aggregate function to be referenced in the select list when
+        database cannot determine that it is functionally dependent on the
+        columns in the `group by`_ clause. See :ref:`Aggregating with MySQL
+        ONLY_FULL_GROUP_BY enabled <aggregation-mysql-only-full-group-by>`.
```
The admonition links back to the aggregation docs.

sarahboyce replied:
```suggestion
        columns in the `group by`_ clause. See the :ref:`aggregation
        documentation <aggregation-mysql-only-full-group-by>` for more details.
```
Possibly

sarahboyce commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,61 @@ aggregate that author count, referencing the annotation field:
 
+.. _aggregation-mysql-only-full-group-by:
+
+Aggregating with MySQL ONLY_FULL_GROUP_BY enabled
+-------------------------------------------------
```
```suggestion
Aggregating with MySQL ``ONLY_FULL_GROUP_BY`` enabled
-----------------------------------------------------
```

I am also tempted to move this section to the bottom of the page as it's quite niche

sarahboyce commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,61 @@ aggregate that author count, referencing the annotation field:
+Aggregating with MySQL ONLY_FULL_GROUP_BY enabled
+-------------------------------------------------
+
+When you are generating an aggregate by using ``.values()`` clause to create the
+grouping for the annotation on MySQL, and the ``ONLY_FULL_GROUP_BY`` SQL mode
+is enabled, it may be necessary to apply :class:`~django.db.models.AnyValue` if
+the annotation includes a mix of aggregate and non-aggregate expressions.
```
```suggestion
When using the ``values()`` clause to group query results for annotations in
MySQL with the ``ONLY_FULL_GROUP_BY`` SQL mode enabled, you may need to apply
:class:`~django.db.models.AnyValue` if the annotation includes a mix of
aggregate and non-aggregate expressions.
```

(with each of these suggestions, only take bits you like)

sarahboyce commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,61 @@ aggregate that author count, referencing the annotation field:
+annotation is composed of aggregate and non-aggregate expressions,
+``num_authors`` is an aggregate expression while ``greatest_page`` isn't.
+
+The column in the group by clause is the expression for ``greatest_pages``.
+MySQL is not able to determine that the ``greatest_pages`` referenced in
+``pages_per_author`` is functionally dependent on the column in the group by
+clause, and it might raise a
+``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
+To avoid the ``OperationalError``, you can wrap the non-aggregate expression
+with :class:`~django.db.models.AnyValue`.
```
```suggestion
Since the grouping is based on the ``greatest_pages`` expression, MySQL may be
unable to determine that ``greatest_pages`` (used in the ``pages_per_author``
expression) is functionally dependent on the grouped column. As a result, it
may raise an error like:

.. code-block:: pytb

    OperationalError: incompatible with sql_mode=only_full_group_by

To avoid this, you can wrap the non-aggregate expression with
:class:`~django.db.models.AnyValue`.
```

sarahboyce commented on docs/topics/db/aggregation.txt:
```diff
@@ -643,6 +643,61 @@ aggregate that author count, referencing the annotation field:
+    ... ).aggregate(Avg("pages_per_author"))
+    {'pages_per_author__avg': 532.57143333}
+
+The other supported databases will not encounter the ``OperationalError`` in the
+example above, because they will detect the functional dependency. In general,
+``AnyValue`` is useful for working around columns in the select list that
+involve non-aggregate functions or complex expressions that are not determined
+by the database to be functionally dependent on the columns in the grouping clause.
```
```suggestion
Other supported databases do not encounter the ``OperationalError`` in the
example above because they can detect the functional dependency. In general,
``AnyValue`` is useful when dealing with select list columns that involve
non-aggregate functions or complex expressions not recognized by the database
as functionally dependent on the columns in the grouping clause.
```

sarahboyce replied:
```suggestion
Other supported databases do not encounter the ``OperationalError`` in the
example above because they can detect the functional dependency. In general,
``AnyValue`` is useful when dealing with select list columns that involve
non-aggregate functions or complex expressions not recognized by the database
as functionally dependent on the columns in the grouping clause.
```

sarahboyce reviewed (commented):
Thank you for the quick updates @ontowhee - those look great!
I think we might want to tighten up the test coverage for the SQLite implementation and I have some docs suggestions (but please let me know if you disagree with any of those)

charettes commented on django/db/backends/sqlite3/_functions.py:
```diff
@@ -513,3 +514,15 @@ class VarPop(ListAggregate):
+
+
+class AnyValue(ListAggregate):
+    def value(self):
+        if self:
+            return self[0]
+        return None
+
+    def finalize(self):
+        if self:
+            return self[0]
+        return None
```
It looks like `value()` was only necessary when we considered adding support for its usage as a window function so it can be entirely removed now.

As for `finalize` it appears that it won't be called on empty groups by SQLite so the `if self` guard can be dropped

```suggestion
    def finalize(self):
        return self[0]
```

The `test_any_value_filter` test already covers it by making sure `None` is returned on empty groups.

ontowhee commented:
I've addressed all the feedback with the following changes:
- Removed `value()` and cleaned up `finalize()` in sqlite3._functions.py
- Updated sub header in the aggregation docs, and moved the entire section to the bottom of the page.
- Applied your suggestions for rewording the documentation. I took most of your suggestions, since they read much better than mine.
- Added the full OperationalError message. It was nice to have the traceback block for this.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce reviewed (approved):
Thank you @ontowhee for all your work on this :star: this looks good to me :tada:

felixxm commented on docs/releases/6.0.txt:
```diff
@@ -212,6 +212,10 @@ Models
 
+* The new :class:`~django.db.models.AnyValue` aggregate returns an arbitrary
+  value from the non-null input values. This is supported on SQLite, MySQL,
+  Oracle, and PostgreSQL 16.0+.
```
```suggestion
  Oracle, and PostgreSQL 16+.
```
